lib/ui/painting.dart

```diff
@@ -1606,14 +1618,150 @@ class Image extends NativeFieldWrapperClass2 {
+      );
+      _disposed = true;
+      return true;
+    }());
+    _dispose();
+
+  }
+
+  void _dispose() native 'Image_dispose';
+
+  /// Returns the native wrapper of the Image. Any calls to `native` must use
+  /// this getter to avoid passing an [_ImageHandle] to native, which will
+  /// crash.
+  Image get _unwrapped => this;
+
+  Set<_ImageHandle> _handles = <_ImageHandle>{};
+
+  /// If asserts are enabled, returns the [StackTrace]s of each open handle from
+  /// [createHandle], in creation order.
+  ///
+  /// If asserts are disabled, this method always returns null.
+  List<StackTrace>? debugGetOpenHandleStackTraces() {
+    assert(!_disposed);
+    List<StackTrace>? stacks;
+    assert(() {
+      stacks = _handles.map((_ImageHandle handle) => handle.debugStack!).toList();
+      return true;
+    }());
+    return stacks;
+  }
+
+  /// Creates a disposable handle to this image.
+  ///
+  /// The returned object behaves identically to this image, except calling
+  /// [dispose] on it will only dispose the underlying native resources if it
+  /// is the last remaining handle.
+  Image createHandle() {
+    if (_disposed) {
+      throw StateError('Object disposed');
+    }
+    final _ImageHandle handle = _ImageHandle(this);
+    _handles.add(handle);
+    return handle;
+  }
 
   @override
   String toString() => '[$width\u00D7$height]';
 }
 
+/// A disposable handle to an [Image].
+///
+/// Handles can create more handles as long as they (and the underlying image)
+/// are not disposed.
+class _ImageHandle implements Image {
+  _ImageHandle(this._image) {
+    assert(() {
+      debugStack = StackTrace.current;
+      return true;
+    }());
+  }
+
+  final Image _image;
+
+  StackTrace? debugStack;
+
+  @override
+  int get width {
+    assert(!_disposed && !_image._disposed);
+    return _image.width;
+  }
+
+  @override
+  int get height {
+    assert(!_disposed && !_image._disposed);
+    return _image.height;
+  }
+
+  @override
+  Future<ByteData?> toByteData({ImageByteFormat format = ImageByteFormat.rawRgba}) {
+    assert(!_disposed && !_image._disposed);
+    return _image.toByteData(format: format);
+  }
+
+  @override
+  Image get _unwrapped {
+    assert(!_disposed && !_image._disposed);
+    return _image;
+  }
+
+  @override
+  List<StackTrace>? debugGetOpenHandleStackTraces() {
+    assert(!_disposed && !_image._disposed);
+    return _image.debugGetOpenHandleStackTraces();
+  }
+
+
+  @override
+  Image createHandle() {
+    assert(!_disposed && !_image._disposed);
+    return _image.createHandle();
+  }
+
+  @override
+  bool _disposed = false;
+
+  @override
+  void dispose() {
+    assert(() {
+      assert(!_disposed && !_image._disposed);
+      assert(_image._handles.contains(this));
+      _disposed = true;
+      return true;
+    }());
+    _image._handles.remove(this);
+    if (_image._handles.isEmpty) {
+      _image.dispose();
+    }
+  }
+
+  /// Unused private implementation of [Image]
```

Comment:
`//` rather than `///` I think.

Reply:
I did this so the link to Image would work in the IDE, but can remove it. I think DartDoc ignores it since there's only private members after it anyway.